### PR TITLE
Part 5 of nullable and modernization

### PR DESCRIPTION
**Pull request by erikmav**
Enable nullable for the whole repo, including enabling some disabled XML tests and adding more test coverage. For #124 .

### CODE REVIEW

File: src/log4net/Util/PatternParser.cs [hunk 1]
```diff
@@ -320,27 +277,28 @@ private void ProcessLiteral(string text)
...
-    private void ProcessConverter(string converterName, string option, FormattingInfo formattingInfo)
+    private void ProcessConverter(string converterName, string? option, FormattingInfo formattingInfo)
     {
-      LogLog.Debug(declaringType, "Converter [" + converterName + "] Option [" + option + "] Format [min=" + formattingInfo.Min + ",max=" + formattingInfo.Max + ",leftAlign=" + formattingInfo.LeftAlign + "]");
+      LogLog.Debug(declaringType, $"Converter [{converterName}] Option [{option}] Format [min={formattingInfo.Min},max={formattingInfo.Max},leftAlign={formattingInfo.LeftAlign}]");
 
       // Lookup the converter type
-      ConverterInfo converterInfo = (ConverterInfo)m_patternConverters[converterName];
-      if (converterInfo == null)
+      ConverterInfo? converterInfo = (ConverterInfo)PatternConverters[converterName];
```
Comment (FreeAndNil) [inline src/log4net/Util/PatternParser.cs:285]: use pattern matching

File: src/log4net/Util/PatternStringConverters/IdentityPatternConverter.cs [hunk 2]
```diff
@@ -43,15 +43,14 @@ internal sealed class IdentityPatternConverter : PatternConverter
...
-    public override void Convert(TextWriter writer, object state)
+    public override void Convert(TextWriter writer, object? state)
     {
       try
       {
-        if (System.Threading.Thread.CurrentPrincipal != null &&
-          System.Threading.Thread.CurrentPrincipal.Identity != null &&
-          System.Threading.Thread.CurrentPrincipal.Identity.Name != null)
+        string? name = System.Threading.Thread.CurrentPrincipal?.Identity?.Name;
+        if (name is not null)
```
Comment (FreeAndNil) [inline src/log4net/Util/PatternStringConverters/IdentityPatternConverter.cs:51]: use pattern matching

File: src/log4net/Util/ThreadContextStacks.cs [hunk 3]
```diff
@@ -24,54 +24,30 @@ namespace log4net.Util
...
     {
       get
       {
-        ThreadContextStack stack = null;
+        ThreadContextStack? stack;
 
-        object propertyValue = m_properties[key];
-        if (propertyValue == null)
+        object? propertyValue = m_properties[key];
+        if (propertyValue is null)
```
Comment (FreeAndNil) [inline src/log4net/Util/ThreadContextStacks.cs:50]: use pattern matching

File: src/log4net/Util/TypeConverters/IPAddressConverter.cs [hunk 4]
```diff
@@ -72,25 +70,22 @@ public bool CanConvertFrom(Type sourceType)
...
             return result;
           }
 
           // Try to resolve via DNS. This is a blocking call. 
           // GetHostEntry works with either an IPAddress string or a host name
-          IPHostEntry host = Dns.GetHostEntry(str);
-          if (host != null &&
-            host.AddressList != null &&
+          IPHostEntry? host = Dns.GetHostEntry(str);
+          if (host?.AddressList is not null &&
```
Comment (FreeAndNil) [inline src/log4net/Util/TypeConverters/IPAddressConverter.cs:86]:
```suggestion
          if (host?.AddressList?.FirstOrDefault() is IPAddress address)
```

**Review by FreeAndNil — CHANGES_REQUESTED** [review]
Thanks a lot.
Just some nits and two things:
- `object[]` (for logging parameters) should become `object?[]?` (otherwise callers have to check them for null values)
- the missing `NotNullIfNotNullAttribute` (is only available for net6 or later, but we can have our own internal version inside log4net)

File: src/log4net/ILog.cs [hunk 5]
```diff
@@ -216,7 +216,7 @@ public interface ILog : ILoggerWrapper
     /// </remarks>
     /// <seealso cref="M:Debug(object)"/>
     /// <seealso cref="IsDebugEnabled"/>
-    void DebugFormat(IFormatProvider provider, string format, params object[] args);
+    void DebugFormat(IFormatProvider? provider, string format, params object[] args);
```
[the remark this answers is not in the document]
Comment (erikmav) [reply src/log4net/ILog.cs:219]: Also propagated new nulls for params like `object arg0` to align with this. Leaving `object message` cases as non-nullable for now, but may need to flip when trying 3.0 binaries in real projects with nullable turned on.

File: src/log4net/Appender/BufferingAppenderSkeleton.cs [hunk 6]
```diff
@@ -437,7 +437,7 @@ protected override void Append(LoggingEvent loggingEvent)
         loggingEvent.Fix = Fix;
 
         // Add to the buffer, returns the event discarded from the buffer if there is no space remaining after the append
-        LoggingEvent discardedLoggingEvent = m_cb.Append(loggingEvent);
+        LoggingEvent? discardedLoggingEvent = m_cb.Append(loggingEvent);
```
[the remark this answers is not in the document]
Comment (FreeAndNil) [reply src/log4net/Appender/BufferingAppenderSkeleton.cs:440]: Alright, I didn't see that.

File: src/log4net/ILog.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (FreeAndNil) [reply src/log4net/ILog.cs:219]: Good idea.
We also have a lot of nullable aware consumers for testing.
Do you need the preview nuget package on nuget.org or can you host it yourself (for testing)?

File: src/log4net/Core/LogImpl.cs [hunk 7]
```diff
@@ -326,7 +329,7 @@ public virtual void DebugFormat(string format, object arg0, object arg1, object
     /// methods instead.
     /// </para>
     /// </remarks>
-    public virtual void DebugFormat(IFormatProvider provider, string format, params object[] args)
+    public virtual void DebugFormat(IFormatProvider? provider, string format, params object[] args)
```
[the remark this answers is not in the document]
Comment (erikmav) [reply src/log4net/Core/LogImpl.cs:332]: Added unit tests for these cases, and ended up with a specific hack for `XxxFormat(message, object? arg0)` not being matched when called with `XxxFormat("{0}", null)` and resulting in LogLog warnings on exceptions trying to render (see below). The overload with `XxxFormat(message, object?[]?)` is matched instead. To ensure `XxxFormat("{0}", null)` results in an expected empty string, the `object?[]?` overload converts a null array to an array of one null, `object?[] { null }`.

log4net:WARN Exception while rendering format
System.FormatException: Index (zero based) must be greater than or equal to zero and less than the size of the argument list.
   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.Format(IFormatProvider provider, String format, Object[] args)
   at log4net.Util.SystemStringFormat.StringFormat(IFormatProvider provider, String format, Object[] args) in C:\log4net2\src\log4net\Util\SystemStringFormat.cs:line 100

**Review by FreeAndNil — APPROVED** [review]
Thank you very much.

**Pull request merged**